src/dump.rs

```diff
@@ -0,0 +1,281 @@
+use crate::{CpuIdReader, CpuIdResult, CpuIdWriter};
+use std::collections::HashMap;
+
+#[derive(Clone)]
+enum LeafOrSubleaves {
+    Leaf(CpuIdResult),
+    Subleaf(HashMap<u32, CpuIdResult>),
+}
+
+// TODO: Clone is necessary because CpuIdReader wants it (for leaves with more complex subleaf
+// structures, like the extended topology info leaf)
+//
+// This implies that there's a full clone of the dump held on for those leaf-specific views, which
+// is unfortunate! It's also not yet really clear how to assemble those more complex leaves for
+// writer purposes.
```

Comment:
`ExtendedTopologyIter` and related are challenging. I *think* they could take a borrow of the reader, which could work for reading and writing as well as for synthetic sources and `CpuIdReaderNative`, but I haven't taken a really close look. But mostly I assume that having a `core::alloc::Vec<CpuIdResult>` of subleaves in the `ExtendedTopologyIter`-like types would be not be okay!